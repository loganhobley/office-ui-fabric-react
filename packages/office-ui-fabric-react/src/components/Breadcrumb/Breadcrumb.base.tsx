import * as React from 'react';
import {
  BaseComponent,
  autobind,
  customizable,
  getRTL
} from '../../Utilities';
import { IFocusZone, FocusZone, FocusZoneDirection } from '../../FocusZone';
import {
  IBreadcrumb,
  IBreadcrumbProps,
  IBreadcrumbItem,
  IBreadcrumbStyles,
  IBreadcrumbStyleProps
} from './Breadcrumb.types';
import { DirectionalHint } from '../../common/DirectionalHint';
import { ResizeGroup } from '../../ResizeGroup';
import { TooltipHost, TooltipOverflowMode } from '../../Tooltip';
import { CrumbBase } from './Crumb.base';
import {
  classNamesFunction
} from '../../Styling';

export interface IBreadCrumbData {
  props: IBreadcrumbProps;
  renderedItems: IBreadcrumbItem[];
  renderedOverflowItems: IBreadcrumbItem[];
}

const OVERFLOW_KEY = 'overflow';
const nullFunction = (): null => null;
const getClassNames = classNamesFunction<IBreadcrumbStyleProps, IBreadcrumbStyles>();

@customizable('Breadcrumb', ['theme'])
export class BreadcrumbBase extends BaseComponent<IBreadcrumbProps, any> implements IBreadcrumb {
  public static defaultProps: IBreadcrumbProps = {
    items: [],
    maxDisplayedItems: 999
  };

  private _classNames: {[key in keyof IBreadcrumbStyles]: string };
  private _focusZone: IFocusZone;

  constructor(props: IBreadcrumbProps) {
    super(props);
  }

  public render() {
    const {
      as,
      onReduceData = this._onReduceData,
      maxDisplayedItems,
      className,
      theme,
      items,
      getStyles
    } = this.props;

    const breadCrumbData: IBreadCrumbData = {
      props: this.props,
      renderedItems: items.slice(-maxDisplayedItems!),
      renderedOverflowItems: items.slice(0, -maxDisplayedItems!)
    };

    this._classNames = getClassNames(getStyles!, { theme: theme!, className });

    return (
      <ResizeGroup
        as={ as }
        className={ this._classNames.root }
        onRenderData={ this._onRenderData }
        onReduceData={ onReduceData }
        data={ breadCrumbData }
      />
    );
  }

  public focus(): void {
    if (this._focusZone) {
      this._focusZone.focus();
    }
  }

  @autobind
  private _onReduceData(data: IBreadCrumbData): IBreadCrumbData | undefined {
    let { renderedItems, renderedOverflowItems } = data;
    let movedItem = renderedItems[0];
    renderedItems = renderedItems.slice(1);

    renderedOverflowItems = [...renderedOverflowItems, movedItem];

    if (movedItem !== undefined) {
      return { ...data, renderedItems, renderedOverflowItems };
    }
  }

  @autobind
  private _onRenderData(data: IBreadCrumbData): JSX.Element {
    let {
      ariaLabel,
      crumbAs: CrumbType = CrumbBase
        } = data.props;
    let classNames = this._classNames;
    let { renderedOverflowItems, renderedItems } = data;

    let contextualItems = renderedOverflowItems.map(
      (item, index) => ({
        name: item.text,
        key: item.key,
        onClick: item.onClick && ((ev: React.MouseEvent<any>) => item.onClick!(ev, item)),
        href: item.href
      })
    );

    return (
      <FocusZone
        className={ classNames.root }
        direction={ FocusZoneDirection.horizontal }
        elementType='ol'
        componentRef={ this._resolveRef('_focusZone') }
        role='navigation'
        aria-label={ ariaLabel }
      >
        { renderedOverflowItems && renderedOverflowItems.length !== 0 && (
          <CrumbType
            key='overflow'
            iconProps={ { iconName: 'More' } }
            menuProps={ {
              items: contextualItems,
              directionalHint: DirectionalHint.bottomLeftEdge
            } }
            withChevron={ renderedItems.length > 0 }
          />
        ) }
        { renderedItems.map(
          (item, index) => (
            <CrumbType
              key={ item.key }
              item={ item }
              withChevron={ index !== (renderedItems.length - 1) }
            />
          )) }
      </FocusZone>
    );
  }
}
