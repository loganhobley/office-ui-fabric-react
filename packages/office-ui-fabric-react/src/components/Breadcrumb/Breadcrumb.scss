@import '../../common/common';
@import 'Breadcrumb.Tokens';

// Copyright (c) Microsoft. All rights reserved. Licensed under the MIT license. See LICENSE in the project root for license information.

//
// Office UI Fabric
// --------------------------------------------------
// Breadcrumb styles

$Breadcrumb-overflowButtonSize: 16px;
$Breadcrumb-itemMaxWidth: 160px;
$Breadcrumb-itemMaxWidth-sm: 116px;

:global {

  .ms-Breadcrumb {
    @include ms-baseFont;
    margin: 23px 0 1px;
  }

  .ms-Breadcrumb-list {
    white-space: nowrap;
    padding: 0;
    margin: 0;
  }

  .ms-Breadcrumb-chevron {
    font-size: $ms-font-size-s;
    color: $breadcrumb-chevron-fg;
    vertical-align: top;
    margin: 11px 7px;
    line-height: 1;
  }

  .ms-Breadcrumb-listItem {
    list-style-type: none;
    vertical-align: top;
    margin: 0;
    padding: 0;
    display: inline-block;
    position: relative;

    &:last-of-type {
      .ms-Breadcrumb-chevron {
        display: none;
      }
    }

    .ms-Breadcrumb-item,
    .ms-Breadcrumb-itemLink {
      @include ms-font-xl;
      color: $breadcrumb-itembutton-fg;
      display: inline-block;
      padding: 0 8px;
      max-width: $Breadcrumb-itemMaxWidth;
      white-space: nowrap;
      text-overflow: ellipsis;
      overflow: hidden;
      vertical-align: top;
    }
  }

  .ms-Breadcrumb-overflow {
    display: inline-block;
    position: relative;
  }

  .ms-Breadcrumb-overflowButton {
    font-size: $Breadcrumb-overflowButtonSize;
    display: inline-block;
    color: $breadcrumb-overflowbutton-fg;
    padding: 8px;
    cursor: pointer;
    vertical-align: top;
  }

  .ms-Breadcrumb-item {
    &:hover {
      cursor: default;
    }
  }

  .ms-Link.ms-Breadcrumb-itemLink {
    &:hover {
      background-color: $breadcrumb-button-hover-bg;
      color: initial;
      cursor: pointer;
    }

    &:focus {
      color: $ms-color-neutralDark;
    }

    &:active {
      outline: transparent;
      background-color: $breadcrumb-button-active-bg;
      color: $breadcrumb-button-active-fg;
    }
  }

  .ms-Breadcrumb-itemLink,
  .ms-Breadcrumb-overflowButton {
    text-decoration: none;
    outline: transparent;
  }

  @media screen and (max-width: $ms-screen-md-max) {
    margin: 11px 0 1px;

    .ms-Breadcrumb-listItem .ms-Breadcrumb-itemLink {
      font-size: $ms-font-size-l;
    }

    .ms-Breadcrumb-chevron {
      font-size: $ms-font-size-mi;
      margin: 8px 5px;
    }

    .ms-Breadcrumb-overflowButton {
      font-size: $ms-font-size-m-plus;
      padding: 5px 4px;
    }
  }

  @media screen and (max-width: $ms-screen-sm-max) {
    .ms-Breadcrumb-listItem .ms-Breadcrumb-itemLink {
      font-size: $ms-font-size-m;
      max-width: $Breadcrumb-itemMaxWidth-sm;
    }

    .ms-Breadcrumb-chevron {
      margin: 5px 4px;
    }

    .ms-Breadcrumb-overflowButton {
      padding: 2px 4px;
    }
  }

}
