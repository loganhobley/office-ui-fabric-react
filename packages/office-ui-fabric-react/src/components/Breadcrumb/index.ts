export * from './Breadcrumb.base';
export * from './Breadcrumb.props';
export * from './Breadcrumb';
export * from './Crumb.props';