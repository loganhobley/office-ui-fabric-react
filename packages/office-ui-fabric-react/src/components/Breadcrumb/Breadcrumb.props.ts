/* tslint:disable:no-unused-variable */
import * as React from 'react';
/* tslint:enable:no-unused-variable */
import { BreadcrumbBase, IBreadCrumbData } from './Breadcrumb.base';
import { IBaseProps, IRenderFunction } from '../../Utilities';
import { ITheme, IStyle, IStyleFunction } from '../../Styling';

export interface IBreadcrumb {
  /**
   * Focuses the active breadcrumb.
   */
  focus(): void;
}

export interface IBreadcrumbProps extends React.HTMLAttributes<HTMLDivElement>, IBaseProps {
  /**
   * Optional callback to access the IBreadcrumb interface. Use this instead of ref for accessing
   * the public methods and properties of the component.
   */
  componentRef?: (component: IBreadcrumb) => void;

  /**
   * Theme (provided through customization.)
   */
  theme?: ITheme;

  /**
   * Collection of breadcrumbs to render
   */
  items: IBreadcrumbItem[];

  /**
   * Optional root classname for the root breadcrumb element.
   */
  className?: string;

  /**
   * The maximum number of breadcrumbs to display before coalescing.
   * If not specified, all breadcrumbs will be rendered.
   */
  maxDisplayedItems?: number;

  /** Method to call when trying to render an item. */
  onRenderItem?: IRenderFunction<IBreadcrumbItem>;

  /**
   * Method to call when reducing the length of the breadcrumb.
   * Return undefined to never reduce breadcrumb length
   */
  onReduceData?: (data: IBreadCrumbData) => IBreadCrumbData | undefined;

  /**
 * Aria label to place on the navigation landmark for breadcrumb
 */
  ariaLabel?: string;

  /**
   * Call to provide customized styling that will layer on top of the variant rules.
   */
  getStyles?: IStyleFunction<IBreadcrumbStyleProps, IBreadcrumbStyles>;

}

export interface IBreadcrumbItem {

  /**
   * Text to display to the user for the breadcrumb
   */
  text: string;

  /**
   * Arbitrary unique string associated with the breadcrumb
   */
  key: string;

  /**
   * Callback issued when the breadcrumb is selected.
   */
  onClick?: (ev?: React.MouseEvent<HTMLElement>, item?: IBreadcrumbItem) => void;

  /**
   * Url to navigate to when this breadcrumb is clicked.
   */
  href?: string;

  /**
   * If this breadcrumb item is the item the user is currently on, if set to true, aria-current="page" will be applied to this breadcrumb link
   */
  isCurrentItem?: boolean;
}

// Given these props:
export interface IBreadcrumbStyleProps {
  theme: ITheme;
  className?: string;
}

// We produce these styles.
export interface IBreadcrumbStyles {
  root: IStyle;
}
