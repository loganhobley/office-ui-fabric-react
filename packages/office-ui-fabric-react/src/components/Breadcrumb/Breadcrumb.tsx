import * as React from 'react';

import {
  IBreadcrumbProps,
  IBreadcrumbStyles,
  IBreadcrumbStyleProps
} from './Breadcrumb.props';
import { BreadcrumbBase } from './Breadcrumb.base';
import { styled } from '../../Styling';
import { ICrumbProps, ICrumbStyleProps, ICrumbStyles } from './Crumb.props';
import { getStyles } from './Breadcrumb.styles';
import { getStyles as getCrumbStyles } from './Crumb.styles';

// Create a Breadcrumb variant which uses these default styles and this styled subcomponent.
export const Breadcrumb = styled(
  BreadcrumbBase,
  getStyles,
  {
    'Crumb': { getStyles: getCrumbStyles }
  }
);
