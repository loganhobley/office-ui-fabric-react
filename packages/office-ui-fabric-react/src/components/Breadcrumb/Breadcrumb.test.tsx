/* tslint:disable:no-unused-variable */
import * as React from 'react';
/* tslint:enable:no-unused-variable */

import * as ReactDOM from 'react-dom';
import * as ReactTestUtils from 'react-dom/test-utils';
import * as renderer from 'react-test-renderer';

import { Breadcrumb } from './Breadcrumb';
import { IBreadcrumbItem, IBreadcrumbProps } from './Breadcrumb.props';

class BreadcrumbWrapper extends React.Component<IBreadcrumbProps, {}> {
  public render(): JSX.Element {
    return <Breadcrumb { ...this.props } />;
  }
}

describe('Breadcrumb', () => {
  it('renders breadcumb correctly', () => {
    const items: IBreadcrumbItem[] = [
      { text: 'TestText1', key: 'TestKey1' },
      { text: 'TestText2', key: 'TestKey2' },
      { text: 'TestText3', key: 'TestKey3' },
      { text: 'TestText4', key: 'TestKey4' }
    ];

    let component = renderer.create(
      <Breadcrumb
        items={ items }
      />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // With overflow
    component = renderer.create(
      <Breadcrumb
        items={ items }
        maxDisplayedItems={ 2 }
      />
    );

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('can call the callback when an item is clicked', () => {
    let callbackValue;
    const clickCallback = (ev: React.MouseEvent<HTMLButtonElement | HTMLAnchorElement>, item: IBreadcrumbItem) => {
      callbackValue = item.key;
    };

    const items: IBreadcrumbItem[] = [
      { text: 'TestText', key: 'TestKey', onClick: clickCallback }
    ];

    let component = ReactTestUtils.renderIntoDocument(
      <BreadcrumbWrapper
        items={ items }
      />
    );

    let renderedDOM = ReactDOM.findDOMNode(component as React.ReactInstance);
    let crumbButton = renderedDOM.querySelector('.ms-Crumb-crumbButton');

    ReactTestUtils.Simulate.click(crumbButton!);
    expect(callbackValue).toEqual('TestKey');
  });

  it('moves items to overflow in the correct order', () => {

    const items: IBreadcrumbItem[] = [
      { text: 'TestText1', key: 'TestKey1' },
      { text: 'TestText2', key: 'TestKey2' },
      { text: 'TestText3', key: 'TestKey3' },
      { text: 'TestText4', key: 'TestKey4' }
    ];

    let component = ReactTestUtils.renderIntoDocument(
      <BreadcrumbWrapper
        items={ items }
        maxDisplayedItems={ 2 }
      />
    );

    let renderedDOM = ReactDOM.findDOMNode(component as React.ReactInstance);
    let crumb = renderedDOM.querySelectorAll('.ms-Crumb-textContent');

    expect(crumb[0].textContent).toEqual('TestText3');
  });

});
