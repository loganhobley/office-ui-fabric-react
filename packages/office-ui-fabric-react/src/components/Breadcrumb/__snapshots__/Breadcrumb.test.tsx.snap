// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb renders breadcumb correctly 1`] = `
<div
  className="ms-ResizeGroup"
>
  <div
    className=""
  >
    <ol
      aria-describedby={undefined}
      aria-label={undefined}
      aria-labelledby={undefined}
      className=
          ms-FocusZone
          ms-Breadcrumb
          {
            align-items: stretch;
            display: flex;
            margin-bottom: 0;
            margin-left: -4px;
            margin-right: -4px;
            margin-top: 0;
            padding-bottom: 0px;
            padding-left: 0px;
            padding-right: 0px;
            padding-top: 0px;
          }
      data-focuszone-id="FocusZone0"
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDownCapture={[Function]}
      role="navigation"
    >
      <li
        className=
            ms-Crumb
            {
              display: flex;
              margin-bottom: 0px;
              margin-left: 0px;
              margin-right: 0px;
              margin-top: 0px;
            }
      >
        <span
          className=
              ms-Crumb-crumbLabel
              {
                padding-bottom: 0;
                padding-left: 4px;
                padding-right: 4px;
                padding-top: 0;
              }
        >
          <div
            aria-describedby={undefined}
            className=
                ms-TooltipHost
                ms-Crumb-textContent
                {
                  -moz-osx-font-smoothing: grayscale;
                  -webkit-font-smoothing: antialiased;
                  font-family: 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                  font-size: 21px;
                  font-weight: 100;
                  height: 40px;
                  line-height: 40px;
                  max-width: 180px;
                  overflow: hidden;
                  text-overflow: ellipsis;
                  vertical-align: top;
                  white-space: nowrap;
                }
                @media screen and (max-width: 639px){& {
                  -moz-osx-font-smoothing: grayscale;
                  -webkit-font-smoothing: antialiased;
                  font-family: 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                  font-size: 17px;
                  font-weight: 300;
                }
                @media screen and (max-width: 479px){& {
                  -moz-osx-font-smoothing: grayscale;
                  -webkit-font-smoothing: antialiased;
                  font-family: 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                  font-size: 14px;
                  font-weight: 400;
                }
            onBlurCapture={[Function]}
            onFocusCapture={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <span>
              TestText1
            </span>
          </div>
        </span>
        <i
          aria-hidden={true}
          aria-label={undefined}
          className=
              ms-Icon
              ms-Crumb-chevron
              {
                display: inline-block;
              }
              {
                color: #666666;
                display: inline-block;
                font-size: 12px;
                height: 40px;
                line-height: 46px;
                overflow: hidden;
                user-select: none;
                vertical-align: top;
              }
              @media screen and (-ms-high-contrast: active){& {
                -ms-high-contrast-adjust: none;
                color: WindowText;
              }
              @media screen and (max-width: 639px){& {
                line-height: 42px;
              }
              @media screen and (max-width: 479px){& {
                font-size: 8px;
              }
          data-icon-name="ChevronRight"
          role="presentation"
        />
      </li>
      <li
        className=
            ms-Crumb
            {
              display: flex;
              margin-bottom: 0px;
              margin-left: 0px;
              margin-right: 0px;
              margin-top: 0px;
            }
      >
        <span
          className=
              ms-Crumb-crumbLabel
              {
                padding-bottom: 0;
                padding-left: 4px;
                padding-right: 4px;
                padding-top: 0;
              }
        >
          <div
            aria-describedby={undefined}
            className=
                ms-TooltipHost
                ms-Crumb-textContent
                {
                  -moz-osx-font-smoothing: grayscale;
                  -webkit-font-smoothing: antialiased;
                  font-family: 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                  font-size: 21px;
                  font-weight: 100;
                  height: 40px;
                  line-height: 40px;
                  max-width: 180px;
                  overflow: hidden;
                  text-overflow: ellipsis;
                  vertical-align: top;
                  white-space: nowrap;
                }
                @media screen and (max-width: 639px){& {
                  -moz-osx-font-smoothing: grayscale;
                  -webkit-font-smoothing: antialiased;
                  font-family: 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                  font-size: 17px;
                  font-weight: 300;
                }
                @media screen and (max-width: 479px){& {
                  -moz-osx-font-smoothing: grayscale;
                  -webkit-font-smoothing: antialiased;
                  font-family: 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                  font-size: 14px;
                  font-weight: 400;
                }
            onBlurCapture={[Function]}
            onFocusCapture={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <span>
              TestText2
            </span>
          </div>
        </span>
        <i
          aria-hidden={true}
          aria-label={undefined}
          className=
              ms-Icon
              ms-Crumb-chevron
              {
                display: inline-block;
              }
              {
                color: #666666;
                display: inline-block;
                font-size: 12px;
                height: 40px;
                line-height: 46px;
                overflow: hidden;
                user-select: none;
                vertical-align: top;
              }
              @media screen and (-ms-high-contrast: active){& {
                -ms-high-contrast-adjust: none;
                color: WindowText;
              }
              @media screen and (max-width: 639px){& {
                line-height: 42px;
              }
              @media screen and (max-width: 479px){& {
                font-size: 8px;
              }
          data-icon-name="ChevronRight"
          role="presentation"
        />
      </li>
      <li
        className=
            ms-Crumb
            {
              display: flex;
              margin-bottom: 0px;
              margin-left: 0px;
              margin-right: 0px;
              margin-top: 0px;
            }
      >
        <span
          className=
              ms-Crumb-crumbLabel
              {
                padding-bottom: 0;
                padding-left: 4px;
                padding-right: 4px;
                padding-top: 0;
              }
        >
          <div
            aria-describedby={undefined}
            className=
                ms-TooltipHost
                ms-Crumb-textContent
                {
                  -moz-osx-font-smoothing: grayscale;
                  -webkit-font-smoothing: antialiased;
                  font-family: 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                  font-size: 21px;
                  font-weight: 100;
                  height: 40px;
                  line-height: 40px;
                  max-width: 180px;
                  overflow: hidden;
                  text-overflow: ellipsis;
                  vertical-align: top;
                  white-space: nowrap;
                }
                @media screen and (max-width: 639px){& {
                  -moz-osx-font-smoothing: grayscale;
                  -webkit-font-smoothing: antialiased;
                  font-family: 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                  font-size: 17px;
                  font-weight: 300;
                }
                @media screen and (max-width: 479px){& {
                  -moz-osx-font-smoothing: grayscale;
                  -webkit-font-smoothing: antialiased;
                  font-family: 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                  font-size: 14px;
                  font-weight: 400;
                }
            onBlurCapture={[Function]}
            onFocusCapture={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <span>
              TestText3
            </span>
          </div>
        </span>
        <i
          aria-hidden={true}
          aria-label={undefined}
          className=
              ms-Icon
              ms-Crumb-chevron
              {
                display: inline-block;
              }
              {
                color: #666666;
                display: inline-block;
                font-size: 12px;
                height: 40px;
                line-height: 46px;
                overflow: hidden;
                user-select: none;
                vertical-align: top;
              }
              @media screen and (-ms-high-contrast: active){& {
                -ms-high-contrast-adjust: none;
                color: WindowText;
              }
              @media screen and (max-width: 639px){& {
                line-height: 42px;
              }
              @media screen and (max-width: 479px){& {
                font-size: 8px;
              }
          data-icon-name="ChevronRight"
          role="presentation"
        />
      </li>
      <li
        className=
            ms-Crumb
            {
              display: flex;
              margin-bottom: 0px;
              margin-left: 0px;
              margin-right: 0px;
              margin-top: 0px;
            }
      >
        <span
          className=
              ms-Crumb-crumbLabel
              {
                padding-bottom: 0;
                padding-left: 4px;
                padding-right: 4px;
                padding-top: 0;
              }
        >
          <div
            aria-describedby={undefined}
            className=
                ms-TooltipHost
                ms-Crumb-textContent
                {
                  -moz-osx-font-smoothing: grayscale;
                  -webkit-font-smoothing: antialiased;
                  font-family: 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                  font-size: 21px;
                  font-weight: 100;
                  height: 40px;
                  line-height: 40px;
                  max-width: 180px;
                  overflow: hidden;
                  text-overflow: ellipsis;
                  vertical-align: top;
                  white-space: nowrap;
                }
                @media screen and (max-width: 639px){& {
                  -moz-osx-font-smoothing: grayscale;
                  -webkit-font-smoothing: antialiased;
                  font-family: 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                  font-size: 17px;
                  font-weight: 300;
                }
                @media screen and (max-width: 479px){& {
                  -moz-osx-font-smoothing: grayscale;
                  -webkit-font-smoothing: antialiased;
                  font-family: 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                  font-size: 14px;
                  font-weight: 400;
                }
            onBlurCapture={[Function]}
            onFocusCapture={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <span>
              TestText4
            </span>
          </div>
        </span>
      </li>
    </ol>
  </div>
</div>
`;

exports[`Breadcrumb renders breadcumb correctly 2`] = `
<div
  className="ms-ResizeGroup"
>
  <div
    className=""
  >
    <ol
      aria-describedby={undefined}
      aria-label={undefined}
      aria-labelledby={undefined}
      className=
          ms-FocusZone
          ms-Breadcrumb
          {
            align-items: stretch;
            display: flex;
            margin-bottom: 0;
            margin-left: -4px;
            margin-right: -4px;
            margin-top: 0;
            padding-bottom: 0px;
            padding-left: 0px;
            padding-right: 0px;
            padding-top: 0px;
          }
      data-focuszone-id="FocusZone5"
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDownCapture={[Function]}
      role="navigation"
    >
      <li
        className=
            ms-Crumb
            {
              display: flex;
              margin-bottom: 0px;
              margin-left: 0px;
              margin-right: 0px;
              margin-top: 0px;
            }
      >
        <button
          aria-current={null}
          aria-describedby={null}
          aria-expanded={false}
          aria-haspopup={true}
          aria-label={undefined}
          aria-labelledby={null}
          aria-owns={null}
          aria-pressed={undefined}
          className=
              ms-Button
              ms-Button--action
              ms-Button--command
              ms-Crumb-crumbButton
              {
                -moz-osx-font-smoothing: grayscale;
                -webkit-font-smoothing: antialiased;
                background-color: transparent;
                border-radius: 0px;
                border: 1px solid transparent;
                box-sizing: border-box;
                color: #333333;
                cursor: pointer;
                display: inline-block;
                font-family: 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                font-size: 14px;
                font-weight: 400;
                height: 40px;
                outline: transparent;
                padding-bottom: 0;
                padding-left: 4px;
                padding-right: 4px;
                padding-top: 0;
                position: relative;
                text-align: left;
                text-decoration: none;
                user-select: none;
                vertical-align: top;
              }
              &::-moz-focus-inner {
                border: 0;
              }
              .ms-Fabric.is-focusVisible &:focus:after {
                border: 1px solid #ffffff;
                bottom: 0px;
                content: "";
                left: 0px;
                outline: 1px solid #666666;
                position: absolute;
                right: 0px;
                top: 0px;
                z-index: 1;
              }
              &:hover {
                background-color: #f4f4f4;
                color: #333333;
                cursor: pointer;
              }
              &:hover .ms-Button-icon {
                color: #0078d7;
              }
              &:active {
                background-color: #c8c8c8;
                color: #333333;
              }
              &:active .ms-Button-icon {
                color: #004578;
              }
          data-is-focusable={true}
          disabled={undefined}
          onClick={[Function]}
          onKeyDown={[Function]}
          type="button"
        >
          <div
            className=
                ms-Button-flexContainer
                {
                  align-items: center;
                  display: flex;
                  flex-wrap: nowrap;
                  height: 100%;
                  justify-content: flex-start;
                }
          >
            <i
              aria-hidden={true}
              aria-label={undefined}
              className=
                  ms-Icon
                  {
                    display: inline-block;
                  }
                  {
                    height: 40px;
                    line-height: 44px;
                  }
              data-icon-name="More"
              role="presentation"
            />
            
          </div>
        </button>
        <i
          aria-hidden={true}
          aria-label={undefined}
          className=
              ms-Icon
              ms-Crumb-chevron
              {
                display: inline-block;
              }
              {
                color: #666666;
                display: inline-block;
                font-size: 12px;
                height: 40px;
                line-height: 46px;
                overflow: hidden;
                user-select: none;
                vertical-align: top;
              }
              @media screen and (-ms-high-contrast: active){& {
                -ms-high-contrast-adjust: none;
                color: WindowText;
              }
              @media screen and (max-width: 639px){& {
                line-height: 42px;
              }
              @media screen and (max-width: 479px){& {
                font-size: 8px;
              }
          data-icon-name="ChevronRight"
          role="presentation"
        />
      </li>
      <li
        className=
            ms-Crumb
            {
              display: flex;
              margin-bottom: 0px;
              margin-left: 0px;
              margin-right: 0px;
              margin-top: 0px;
            }
      >
        <span
          className=
              ms-Crumb-crumbLabel
              {
                padding-bottom: 0;
                padding-left: 4px;
                padding-right: 4px;
                padding-top: 0;
              }
        >
          <div
            aria-describedby={undefined}
            className=
                ms-TooltipHost
                ms-Crumb-textContent
                {
                  -moz-osx-font-smoothing: grayscale;
                  -webkit-font-smoothing: antialiased;
                  font-family: 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                  font-size: 21px;
                  font-weight: 100;
                  height: 40px;
                  line-height: 40px;
                  max-width: 180px;
                  overflow: hidden;
                  text-overflow: ellipsis;
                  vertical-align: top;
                  white-space: nowrap;
                }
                @media screen and (max-width: 639px){& {
                  -moz-osx-font-smoothing: grayscale;
                  -webkit-font-smoothing: antialiased;
                  font-family: 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                  font-size: 17px;
                  font-weight: 300;
                }
                @media screen and (max-width: 479px){& {
                  -moz-osx-font-smoothing: grayscale;
                  -webkit-font-smoothing: antialiased;
                  font-family: 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                  font-size: 14px;
                  font-weight: 400;
                }
            onBlurCapture={[Function]}
            onFocusCapture={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <span>
              TestText3
            </span>
          </div>
        </span>
        <i
          aria-hidden={true}
          aria-label={undefined}
          className=
              ms-Icon
              ms-Crumb-chevron
              {
                display: inline-block;
              }
              {
                color: #666666;
                display: inline-block;
                font-size: 12px;
                height: 40px;
                line-height: 46px;
                overflow: hidden;
                user-select: none;
                vertical-align: top;
              }
              @media screen and (-ms-high-contrast: active){& {
                -ms-high-contrast-adjust: none;
                color: WindowText;
              }
              @media screen and (max-width: 639px){& {
                line-height: 42px;
              }
              @media screen and (max-width: 479px){& {
                font-size: 8px;
              }
          data-icon-name="ChevronRight"
          role="presentation"
        />
      </li>
      <li
        className=
            ms-Crumb
            {
              display: flex;
              margin-bottom: 0px;
              margin-left: 0px;
              margin-right: 0px;
              margin-top: 0px;
            }
      >
        <span
          className=
              ms-Crumb-crumbLabel
              {
                padding-bottom: 0;
                padding-left: 4px;
                padding-right: 4px;
                padding-top: 0;
              }
        >
          <div
            aria-describedby={undefined}
            className=
                ms-TooltipHost
                ms-Crumb-textContent
                {
                  -moz-osx-font-smoothing: grayscale;
                  -webkit-font-smoothing: antialiased;
                  font-family: 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                  font-size: 21px;
                  font-weight: 100;
                  height: 40px;
                  line-height: 40px;
                  max-width: 180px;
                  overflow: hidden;
                  text-overflow: ellipsis;
                  vertical-align: top;
                  white-space: nowrap;
                }
                @media screen and (max-width: 639px){& {
                  -moz-osx-font-smoothing: grayscale;
                  -webkit-font-smoothing: antialiased;
                  font-family: 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                  font-size: 17px;
                  font-weight: 300;
                }
                @media screen and (max-width: 479px){& {
                  -moz-osx-font-smoothing: grayscale;
                  -webkit-font-smoothing: antialiased;
                  font-family: 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
                  font-size: 14px;
                  font-weight: 400;
                }
            onBlurCapture={[Function]}
            onFocusCapture={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <span>
              TestText4
            </span>
          </div>
        </span>
      </li>
    </ol>
  </div>
</div>
`;
